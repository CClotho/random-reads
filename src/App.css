#root {
  max-width: 1280px; /* remove if full width*/
  margin: 0 auto;
  padding: 2rem;
  text-align: center;
  
}

ul, li {
 list-style-type: none;
}


