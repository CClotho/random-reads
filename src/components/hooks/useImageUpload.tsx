import { useEffect, useState } from "react";

interface ImageUpload {
    selectedFile: Blob | MediaSource | null,
    preview: | string | undefined ,
    //setSelectedFile: React.Dispatch<React.SetStateAction<Blob>>;
    setPreview: React.Dispatch<React.SetStateAction<string | undefined>>;
    onSelectedFile: (e: React.ChangeEvent<HTMLInputElement>)=> void;
}

const useImageUpload = function(file?: Blob): ImageUpload {
    
    const [selectedFile, setSelectedFile] = useState<Blob | MediaSource |null>(file ? file : null);
    const [preview, setPreview] = useState<string | undefined >('');

    useEffect(() => {
        
        // If there's no new selected image/file  still set the current preview state.
        if(!selectedFile ) {
            return;
        }

     

        const objectUrl = URL.createObjectURL(selectedFile)
        setPreview(objectUrl)

        return () => URL.revokeObjectURL(objectUrl)
    }, [selectedFile])
    
    const onSelectedFile = function (e: React.ChangeEvent<HTMLInputElement>): void {
        e.stopPropagation();
       
        if (e.target.files && e.target.files.length > 0) {
            const file = e.target.files[0]; // Get the first selected file
            if (file instanceof Blob && file !== selectedFile ) {
              setSelectedFile(file); // Set the valid Blob (File) to selectedFile
            }
          } else {
            setSelectedFile(null); // Reset to null if no file is selected
          }
    }
    
    return {
        selectedFile,
        preview,
        onSelectedFile,
        setPreview
    }
}

export default useImageUpload;