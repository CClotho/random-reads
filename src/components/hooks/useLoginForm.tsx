
interface Login {
    
}

const useLoginForm = function() {
    
    
    return {

    }
}


export default useLoginForm;