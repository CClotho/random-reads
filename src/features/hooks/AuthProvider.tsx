import { Login } from "features/api/authentication";
import { useEffect, useState , createContext, useContext} from "react";
import { fetchUser } from "features/api/user";

// provide the data across the application using React Context Provider
// then store it in localStorage and get the tokens for every request to the server

// check if accessToken exist then proceed otherwise get refresh Token

 const AuthContext = createContext<any>(null);

export const useAuth = () => useContext(AuthContext);


const AuthProvider = function({children}: {children:React.ReactNode}): React.JSX.Element {

   const[loading, setLoading]  = useState(false);
    const [authentication, setAuth] = useState(false); 
     const [user, setUser] = useState<any>({});
    const access = localStorage.access?  localStorage.getItem('access') : '';

         useEffect(()=> {
            console.log("This runs everytime you call Auth provider on a component which is logically to add the refetch of refreshTokens here")
           const fetchData = async function()  {
             console.log(access)
                if(access) {
                
                 setAuth(true)
                 console.log("User Authenticated Status:", authentication)
                 try {
                  const response: any = await fetchUser();
                  setUser(response);
                  setAuth(true);
                  console.log(response);
                  
                } catch (error) {
                  console.error('Error fetching user:', error);
                } finally {
                  setLoading(false);
                }
                 // The refresh route in dummyjson is not available only login and user/me
                /*  fetch('https://dummyjson.com/user/refresh', {
                    method: 'POST',
                    headers: {
                       'Content-Type': 'application/json',
                    },
                    body: JSON.stringify({refreshToken: localStorage.refresh})
  
                 })
                 .then(res => res.json())
                 .then((data)=> {
                    setAuth(true);
                    localStorage.setItem('access', data.accessToken);
                    localStorage.setItem('refresh', data.refreshToken); 
                    console.log("Refreshed", data);
                 }); */
             
           }
           else if(!access) {
              console.log("User Authenticated Status:", authentication)
              console.log("User is logged out.")
              localStorage.clear();
              setAuth(false);
           }
           setLoading(false);
        }
           fetchData();
      },[access, authentication])
      

     const context = {
        loading,
        user,
        authentication,
        setLoading,
        setAuth,
        setUser
     }
    

     return (
      <AuthContext.Provider value={context}>
          {loading && access ? <p>Loading... from AuthProvider</p> : children}
      </AuthContext.Provider>
     )

}

export default AuthProvider;