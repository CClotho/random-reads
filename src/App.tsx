
import './App.css'
import { Link, Outlet } from 'react-router-dom';
import LoginForm from '@components/login-feature/LoginForm';
import SignUpForm from '@components/signup-feature/SignUpForm';
import Avatar from '@components/common/Avatar';

function App() {


  return (
    
    <>
     {/* Not working*/}
     <div>
     <Link to="/story"> Go to story page</Link>
     <Link to="/"> Go to new</Link>
     </div>
      <SignUpForm/>
      <LoginForm/>
      <Outlet />
    </>
  )
}

export default App;
