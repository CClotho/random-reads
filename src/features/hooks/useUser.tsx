import React, { useEffect,useState , createContext, useContext} from "react";
import { fetchUser } from "features/api/user";
import { useAuth } from "./AuthProvider";

const UserContext = createContext<any>(null);
export const useUser=() => useContext(UserContext);
// Simple caching of user's basic information to render for UI
// Otherwise subsequent request will be implemented if we only store it on a state because state (memory) resets everytime
const UserProvider = function({children}: {children: React.ReactNode}): React.JSX.Element {
   const firstName = localStorage.getItem('firstName');
    const lastName = localStorage.getItem('lastName');
    const User = {firstName, lastName};
    const [loading, setLoading] = useState<boolean>(false);
    const [user, setUser] = useState<any>(User);
    const {setAuth, authentication} = useAuth();
    const access = localStorage.access?  localStorage.getItem('access') : '';

    useEffect(()=> {
        const fetchProfile =async function(): Promise<any> {
           // will only fetch if there either local storage values is present
            if (access && authentication && (!firstName || !lastName)) {
                setLoading(true);
                console.log("It rerendered")
                console.log("Current state of user before fetching", user)
        
                try {
                  const response: any = await fetchUser();
                  localStorage.setItem('firstName', response.firstName);
                  localStorage.setItem('lastName', response.lastName);
                  setUser(response);
                  setAuth(true);
                  console.log(response);
                  
                } catch (error) {
                  console.error('Error fetching user:', error);
                } finally {
                  setLoading(false);
                }
              }
            else if(!authentication) {
              console.log("User is logged out.")
              localStorage.clear();
              setAuth(false);
              
              }
          };
            
        fetchProfile();
         
    }, [access, firstName, lastName])  
    const context = {
        user,
        loading,
        setUser,
        setLoading
    }
       
    return (
        <UserContext.Provider value={context}>
        {loading && access ? <p>Loading...from UserProvider</p> : children}
       </UserContext.Provider>
    )
    
}

export default UserProvider;