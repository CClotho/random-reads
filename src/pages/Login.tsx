import LoginForm from "@components/login-feature/LoginForm";
import { useAuth } from 'features/hooks/AuthProvider';
import { redirect } from 'react-router-dom';
import { useEffect } from "react";

const Login = function(): React.JSX.Element {
    
    const {authentication} = useAuth();
 
   
  
    if(!authentication) {
      redirect("/login")
    }
 
    
    return (

        <>
            <nav></nav>
            <div className="form-wrapper">
                <LoginForm/>
            
            </div>
        </>
    )
}


export default Login;