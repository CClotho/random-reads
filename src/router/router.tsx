import { AuthContext, AuthContextType, useAuth } from "features/hooks/AuthProvider";
import {routes} from "./routes";
import { createBrowserRouter,  RouterProvider, Memo, createMemoryRouter } from "react-router-dom";


import Login from "@pages/Login";
import App from "../App";
import Home from "../pages/Home";
import Post from "@components/story-feature/Post";
import SignUp from "@pages/SignUp";
import MainPost from "@components/story-feature/MainPost";
import Stories from "@pages/Stories";
import ProtectedRoute from "@pages/ProtectedRoute";
import Profile from "@pages/Profile";

const router = createBrowserRouter(routes)

export default  router;
const memory_routes =  [

    {  
      /* Layout Element only */
      element: <App />,
      children: [
          { path: "/", element: <Home /> },
         
          { element: <ProtectedRoute />,
            children: [
                { 
                  path: "/stories",
                  element: <Stories />,
                  children: [
                    { path: "/stories/:id", element: <MainPost /> }
                    ]
                },
                {path: "/profile", element: <Profile/>},
              ]
          }
      ]
    },
    {   path: '/login',
        element: <Login/>
        
    },
    {   path: '/sign-up',
        element: <SignUp/>
        
    }
   /*  {path:"/stories/:id" ,element: <MainPost/>} */

 ];

const memory_router = createMemoryRouter(memory_routes, {
    initialEntries: 
    ["/profile"], // Start from home page where user is logged in
    initialIndex: 0}
)


export const AppRouter = function(context : AuthContextType ):React.JSX.Element {

  
    return (
        <AuthContext.Provider value={context}>
            <RouterProvider router={router} />
        </AuthContext.Provider>
    )
}


export const MemoryApp2 = function({context} :{context: AuthContextType} ):React.JSX.Element {

  
    return (
        <AuthContext.Provider value={context}>
            <RouterProvider router={memory_router} />
        </AuthContext.Provider>
    )
}






//Should use this because it make sure we're using the right values that is returned from our Authprovider

/* 
export const AppRouter2 = function():React.JSX.Element {
        
    const context = useAuth();
    return (
        <AuthContext.Provider value={context}>
            <RouterProvider router={router} />
        </AuthContext.Provider>
    )
}
 */