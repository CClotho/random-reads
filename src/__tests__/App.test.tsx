import { vi, Mock,describe, it, expect} from "vitest";
import { render, screen, waitFor, waitForElementToBeRemoved} from "@testing-library/react";
import '@testing-library/jest-dom'; // npm i -D @types/testing-library__jest-dom
import App from "../App";
import Login from "@pages/Login";
import { BrowserRouter, Link, createMemoryRouter } from "react-router-dom";
import {  User, AuthContextType } from "features/hooks/AuthProvider";
import React from 'react'
import { AppRouter, MemoryApp, MemoryApp2 } from "router/router";
import userEvent from '@testing-library/user-event';



/* 
vi.mock("@components/layout/NavBar", ()=> {
    return {
        default: () => (
            <nav>
                <ul>
                <li className="links"><Link to="/test"></Link> Test</li>
            </ul>
            </nav>
        )
    }
}) */


export interface MockType extends AuthContextType  {
    loading: boolean,
    authentication: boolean,
    user: User | null,
    setLoading: Mock
    setAuth: Mock,
    setUser: Mock,
    };

    const Users : AuthContextType ={
        loading: true,
        authentication: true,
        user: {firstName: "Krosi", lastName:"Krey"},
        setLoading: vi.fn(),
        setAuth: vi.fn(),
        setUser: vi.fn(),
    };

      const Users2: AuthContextType = {
        loading: true,
        authentication: false,
        user: { firstName: "Krosi", lastName: "Krey" },
        setLoading: vi.fn(),
        setAuth: vi.fn() as React.Dispatch<React.SetStateAction<boolean>>,
        setUser: vi.fn() as React. Dispatch<React.SetStateAction<User | null>>,
    };

const publicUsers: MockType= {
    loading: false, // problem i set this state to be true so it stuck to true loading state
    authentication: false,
    user: {firstName: "", lastName: ""},
    setLoading:  vi.fn((e: boolean)=> {publicUsers.loading = e; }),
    setAuth: vi.fn(),
    setUser: (vi.fn()),
  };
  
  





  describe('Main App renders the routes', () => {
        it('It should render the public navbar for public users', () => {
            // render(<BrowserRouter><AuthContext.Provider value={publicUsers}><App/></AuthContext.Provider></BrowserRouter>) 
            render(<AppRouter {...publicUsers}/>)

            const NavBar = screen.getByRole('navigation');
            const Links = screen.getAllByRole('listitem')
            expect(NavBar).toBeInTheDocument();
            expect(Links.length).toEqual(5)

                                
            })
            
            //'should not show profile page'
        it('Public users should not see profile page', async ()=> {

    
           // render(<AppRouter {...publicUsers}/>)
           const {rerender} = render(<MemoryApp2 context={publicUsers} />)
           
         
           
            
            const expectedLinks = ['Stories', 'Genre', 'Home', 'Sign Up', 'Login'];
            const listItems = screen.getAllByRole('listitem'); // Get all list items
          
                expectedLinks.forEach((linkText) => {
                // Check if listItems is an array otherwise convert the data to array
                const linkItem = Array.from(listItems).find((item) => {
                return item.textContent ?  item.textContent.trim() === linkText : '';
                });

                    if (linkItem) {
                    expect(linkItem).toBeInTheDocument();
                    expect(linkItem).toMatchSnapshot();
                    } else {
                        console.warn(`Link not found for: ${linkText}`); // Log a warning if not found
                    }
                })
                const heading = screen.getByRole('heading', { level:1, name: /short stories/i }); // or getByText if it's not a role-based heading
                expect(heading).toBeInTheDocument();
                
             /*    const heading = screen.queryByRole('heading', { level: 1 });
                expect(heading).not.toBeInTheDocument() */
                screen.debug();
               
         })

         
        

       
        it('Should redirect user to login form ', async ()=> {
            // check if there is  a user logged in if it navigate to login ro
            //it should show homepage
            // or it should not show the loginpage
                 const user = userEvent.setup()
             /*     const routes = [
                     { path: '/', element: <App/>},
                     {path: '/login', element: <Login/>}
                    
                ]
                 const router = createMemoryRouter(routes) */

                render(<MemoryApp2 context={publicUsers} />) // both works fine problem is when I set authentication to true or false
                //render(<AppRouter {...publicUsers}/>)
               // expect(publicUsers.authentication).toBe(true); 
                const NavBar = screen.getByRole('navigation');
                expect(NavBar).toHaveClass('nav-bar');
                const loginLink = screen.getByRole('link', {name: 'Login'})
                expect(loginLink).toBeInTheDocument()
                 screen.debug();
                await user.click(loginLink);
                expect(screen.getByRole('button', {name: /Login/i})).toBeInTheDocument()
                

        }) 
    

}) 
 
// This wont work unless I comment out these test above^




/*    
  
        
const NavBar = screen.getByRole('navigation');
        const Links = screen.getAllByRole('listitem')
        Links.forEach((link)=> {
            const liText = link.textContent;
            if(liText && liText.trim().length > 0) {
        
                const checkLi = screen.getByText(liText);
                expect(checkLi).toBeInTheDocument();
                expect(checkLi).toMatchSnapshot();
            }
            else {
                expect(link).toBeEmptyDOMElement();
            }
    
        }) */