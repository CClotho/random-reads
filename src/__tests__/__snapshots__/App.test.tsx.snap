// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Main App renders the routes > Public users should not see profile page 1`] = `
<li
  class="links"
>
  <a
    href="/stories"
  >
    Stories
  </a>
</li>
`;

exports[`Main App renders the routes > Public users should not see profile page 2`] = `
<li
  class="links"
>
  <a
    href="/genre"
  >
    Genre
  </a>
</li>
`;

exports[`Main App renders the routes > Public users should not see profile page 3`] = `
<li
  class="links"
>
  <a
    href="/"
  >
    Home
  </a>
</li>
`;

exports[`Main App renders the routes > Public users should not see profile page 4`] = `
<li
  class="links"
>
  <a
    href="/sign-up"
  >
    Sign Up
  </a>
</li>
`;

exports[`Main App renders the routes > Public users should not see profile page 5`] = `
<li
  class="links"
>
  <a
    href="/login"
  >
    Login
  </a>
</li>
`;
